### PR TITLE
string2bytes without  mem-allocs

### PR DESCRIPTION
**Pull request by tylitianrui**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by ptabor** [comment]
```
% (cd pkg && 'go' 'vet' './...')
stderr: # go.etcd.io/etcd/pkg/v3/stringutil
stderr: vet: stringutil/convert_test.go:21:25: cannot use i (variable of type int) as uint value in argument to randString
```

**Comment by AlexStocks** [comment]
how about the benchmark result of "const str = "hello, world""?

**Comment by AlexStocks** [comment]
if the StringToBytes is implemented as above, the test result is as follows.

// const str = "hello, world" // 11Bytes
goos: darwin
goarch: amd64
pkg: github.com/dubbogo/gost/strings
BenchmarkStringToBytes
BenchmarkStringToBytes-12    	1000000000	         0.279 ns/op

// const str = "hello, world" // 11Bytes
goos: darwin
goarch: amd64
pkg: github.com/dubbogo/gost/strings
BenchmarkStringToBytesRaw
BenchmarkStringToBytesRaw-12    	1000000000	         0.253 ns/op

// const str = "hello, world, hello, world, helL" // 24 Bytes
goos: darwin
goarch: amd64
pkg: github.com/dubbogo/gost/strings
BenchmarkStringToBytes
BenchmarkStringToBytes-12    	1000000000	         0.256 ns/op

// const str = "hello, world, hello, world, helL" // 24 Bytes
goos: darwin
goarch: amd64
pkg: github.com/dubbogo/gost/strings
BenchmarkStringToBytesRaw
BenchmarkStringToBytesRaw-12    	1000000000	         0.253 ns/op

const str = "hello, world, hello, world, hello, world" // 40 Bytes
goos: darwin
goarch: amd64
pkg: github.com/dubbogo/gost/strings
BenchmarkStringToBytes
BenchmarkStringToBytes-12    	1000000000	         0.254 ns/op

const str = "hello, world, hello, world, hello, world" // 40 Bytes
goos: darwin
goarch: amd64
pkg: github.com/dubbogo/gost/strings
BenchmarkStringToBytesRaw
BenchmarkStringToBytesRaw-12    	1000000000	         0.269 ns/op

// const str = "dfjaslfjaldsfjalsdfjdl;fjadlfjal;sdfjl;adsfjladsfkadssdfadsfadsf" // 64bytes
goos: darwin
goarch: amd64
pkg: github.com/dubbogo/gost/strings
BenchmarkStringToBytes
BenchmarkStringToBytes-12    	1000000000	         0.280 ns/op

// const str = "dfjaslfjaldsfjalsdfjdl;fjadlfjal;sdfjl;adsfjladsfkadssdfadsfadsf" // 64bytes
goos: darwin
goarch: amd64
pkg: github.com/dubbogo/gost/strings
BenchmarkStringToBytesRaw
BenchmarkStringToBytesRaw-12    	39803451	        29.9 ns/op

the summary is:

if the string length is less than 24 bytes, the shorter the string, the better the performance of `[]byte(string)`.
if the string length is greater than 24 bytes, the longer the string, the better the performance of `StringToBytes`.

**Comment by AlexStocks** [comment]
if u really wanna use this tricky, the implementation should be as follows.

type sbytes struct {
	string
	Cap int
}

// StringToBytes converts a string into bytes without mem-allocs.
func StringToBytes(s string) []byte {
         if len(s) < 24 {
             return []byte(s)
        }

	return *(*[]byte)(unsafe.Pointer(&sbytes{s, len(s)}))
}

File: pkg/stringutil/convert.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package stringutil
+
+import (
+	"unsafe"
+)
+
+// StringToBytes convert a string into bytes without mem-allocs.
+func StringToBytes(s string) []byte {
```
[the remark this answers is not in the document]
Comment (AlexStocks) [reply pkg/stringutil/convert.go:8]: see my comments in the coversation. can not agree with u more.

**Comment by tylitianrui** [comment]
Thanks for the details. I agree with you. Let's close this.

**Pull request closed without merging**